File: 02x10/c-015/main.rb
def traverseGrid x, y, mapx
    if x == 1 or y == 1 then
        return 1
    end
    if !mapx.has_key? x then
        mapx[x] = {}
    end
    mapy = mapx[x]
    if mapy.has_key? y then
        return mapy[y]
    end
    if x == y then
        result = traverseGrid(x, y-1, mapx) * 2
        mapy[y] = result
        return result
    end
    result = traverseGrid(x-1, y, mapx) + traverseGrid(x, y-1, mapx)
    mapy[y] = result
    result
end

def routesGrid grid
    mapx = {}
    corner = grid + 1
    totalRoutes = traverseGrid(corner, corner, mapx)
    totalRoutes
end

puts routesGrid(20)

File: 02x10/c-015/main.php
<?php
function traverseGrid($x, $y, &$mapx) {
    if ($x == 1 || $y == 1) {
        return 1;
    }
    if (!array_key_exists($x, $mapx)) {
        $mapx[$x] = [];
    }
    $mapy = $mapx[$x];
    if (array_key_exists($y, $mapy)) {
        return $mapy[$y];
    }
    if ($x == $y) {
        $result = traverseGrid($x, $y-1, $mapx) * 2;
        $mapx[$x][$y] = $result;
        return $result;
    }
    $result = traverseGrid($x-1, $y, $mapx) + traverseGrid($x, $y-1, $mapx);
    $mapx[$x][$y] = $result;
    return $result;
}

function routesGrid($grid) {
    $mapx = [];
    $corner = $grid + 1;
    $totalRoutes = traverseGrid($corner, $corner, $mapx);
    return $totalRoutes;
}

print routesGrid(20);
